### PR TITLE
Color features

### PR DESCRIPTION
**Pull request by jsarha**
Contains test for YCbCr  and some feature

### CODE REVIEW

File: kms++util/inc/kms++util/color.h [hunk 1]
```diff
@@ -29,9 +37,12 @@ struct RGB
 
 struct YUV
 {
+	static const int coef[static_cast<unsigned>(YUVType::MAX)][3][3];
+	static const int offset[static_cast<unsigned>(YUVType::MAX)][3];
```
Comment (tomba) [inline kms++util/inc/kms++util/color.h]: These seem to be extra

**Comment by jsarha** [comment]
On 05/26/17 13:23, Tomi Valkeinen wrote:
>
> *@tomba* commented on this pull request.
>
> ------------------------------------------------------------------------
>
> In kms++util/inc/kms++util/color.h
> <https://github.com/tomba/kmsxx/pull/23#discussion_r118674757>:
>
> > @@ -29,9 +37,12 @@ struct RGB
>  
>  struct YUV
>  {
> +	static const int coef[static_cast<unsigned>(YUVType::MAX)][3][3];
> +	static const int offset[static_cast<unsigned>(YUVType::MAX)][3];
>
> These seem to be extra
>

Maybe extra, but needed. The static coef tables are defined in
kms++util/
<https://github.com/tomba/kmsxx/pull/23#discussion_r118674757>src/color.c,
and used in the conversions.

**Comment by tomba** [comment]
On 29/05/17 23:15, jsarha wrote:
> On 05/26/17 13:23, Tomi Valkeinen wrote:
>>
>> *@tomba* commented on this pull request.
>>
>> ------------------------------------------------------------------------
>>
>> In kms++util/inc/kms++util/color.h
>> <https://github.com/tomba/kmsxx/pull/23#discussion_r118674757>:
>>
>> > @@ -29,9 +37,12 @@ struct RGB
>>
>> struct YUV
>> {
>> + static const int coef[static_cast<unsigned>(YUVType::MAX)][3][3];
>> + static const int offset[static_cast<unsigned>(YUVType::MAX)][3];
>>
>> These seem to be extra
>>
> 
> Maybe extra, but needed. The static coef tables are defined in
> kms++util/
> <https://github.com/tomba/kmsxx/pull/23#discussion_r118674757>src/color.c,
> and used in the conversions.

Yes, used in color.cpp, but nowhere else, right? Usually declaring
something static in a header file makes no sense, as the definition,
wherever it is, is not visible outside the compilation unit.

 Tomi

**Comment by jsarha** [comment]
On 05/30/17 08:40, Tomi Valkeinen wrote:
> Yes, used in color.cpp, but nowhere else, right? Usually declaring
> something static in a header file makes no sense, as the definition,
> wherever it is, is not visible outside the compilation unit.
> 

Yes, but the coefs are accessible from outside YUV class trough the
static declarations, but I guess there is no real use for them.

**Comment by tomba** [comment]
On 30/05/17 13:11, jsarha wrote:
> On 05/30/17 08:40, Tomi Valkeinen wrote:
>> Yes, used in color.cpp, but nowhere else, right? Usually declaring
>> something static in a header file makes no sense, as the definition,
>> wherever it is, is not visible outside the compilation unit.
>>
> 
> Yes, but the coefs are accessible from outside YUV class trough the
> static declarations, but I guess there is no real use for them.

Oh, I see. You mean you need them in the header file so that color.cpp
can access them?

I think it's better to not have them in the header file, as that makes
them public (even if they can't be accessed anywhere else...) and just
use normal static const variables in color.cpp. You don't have to have
them inside YUV class, if that's causing problems.

 Tomi

**Comment by jsarha** [comment]
On 05/30/17 13:41, Tomi Valkeinen wrote:
>> Yes, but the coefs are accessible from outside YUV class trough the
>> static declarations, but I guess there is no real use for them.
> 
> Oh, I see. You mean you need them in the header file so that color.cpp
> can access them?
> 
> I think it's better to not have them in the header file, as that makes
> them public (even if they can't be accessed anywhere else...) and just
> use normal static const variables in color.cpp. You don't have to have
> them inside YUV class, if that's causing problems.

Of course they can be accessed outside. They are public members of the
YUV class. Static for a class member is a bit different thing than
static for a global variable.

But then again I see no need good reason to anybody to access them
outside, so they can be taken away from the header file in any case.

**Comment by tomba** [comment]
> Of course they can be accessed outside. They are public members of the
> YUV class. Static for a class member is a bit different thing than
> static for a global variable.

Of course... Sorry my c++ cluelesness =). But yes, I agree, it's unlikely that they'll be used anywhere else, so the .cpp file is a better place for them for now.

**Comment by jsarha** [comment]
A new pull request version. Changes:
- remove "back to default" step from plane_csc.py
- update ctm_test.py prints to match what should be seen on the screen
- remove public YUV coefficients and offsets from YUV class

**Comment by tomba** [comment]
Use chmod a+x for python scripts so that they can be executed

File: py/tests/ctm_test.py [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+def ctm_to_blob(ctm, card):
+    len=9
+    arr = bytearray(len*8)
+    view = memoryview(arr).cast("I")
+
+    for x in range(len):
+        i, d = divmod(ctm[x], 1)
+        if i < 0:
+            i = -i
+            sing = 1 << 31
```
Comment (tomba) [inline py/tests/ctm_test.py]: No singing please!

**Comment by jsarha** [comment]
On 05/31/17 10:48, Tomi Valkeinen wrote:
> +card = pykms.Card()
> +res = pykms.ResourceManager(card)
> +conn = res.reserve_connector("DPI-1")
> 
> I think it's better to leave the connector name out, unless there's a
> specific reason why only dpi-1 works. Or if you want to add a few lines,
> you can get the conn name from cmdline (see test.py)
> 


The primary reason is that CPR does not have any effect on HDMI output.
But I guess the default output should always be on LCD if it is
available. I'll change this to default output and add a command line
option to change it.

File: py/tests/plane_csc.py [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+yuv_types.append(pykms.YUVType.BT709_Lim)
+yuv_types.append(pykms.YUVType.BT709_Full)
+
+for i in range(0, 2):
+    for j in range(0, 2):
+        input("press enter to for encoding %d range %d\n" % (i, j))
+
+        plane.set_prop("COLOR_ENCODING", i);
+        plane.set_prop("COLOR_RANGE", j);
+
```
Comment (tomba) [inline py/tests/plane_csc.py]: Probably doesn't matter much, but I think you should use atomic commit here, with the two properties set at once.

File: py/tests/plane_csc.py [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+                "zorder": 0})
+r = req.commit_sync()
+print("Plane enable request returned %d\n" % r)
+
+yuv_types = []
+yuv_types.append(pykms.YUVType.BT601_Lim)
+yuv_types.append(pykms.YUVType.BT601_Full)
+yuv_types.append(pykms.YUVType.BT709_Lim)
+yuv_types.append(pykms.YUVType.BT709_Full)
+
```
Comment (tomba) [inline py/tests/plane_csc.py]: Add those values to yuv_types in the array initializer.

**Comment by jsarha** [comment]
Forget this update... bad squash. I'll reorder an push again soon.

**Comment by jsarha** [comment]
Changes:
- ctm_test.py
  - No singing
  - use default output 
  - add command line parameter for selecting output
  - change printing order
- Add DrmPropObject.get_prop() and Property.enums to pykmsbase.cpp
- plane_csc.py
  - direct initialization of yuv_types
  - use enum names in prints
  - use atomic commit for setting endcoding and range

**Pull request merged**